pkg/storagedistribution/storagedistribution.go

```diff
@@ -0,0 +1,390 @@
+func GetStorageUpdateConfig(
+	request *cloudops.StoragePoolUpdateRequest,
+	decisionMatrix *cloudops.StorageDecisionMatrix,
+) (*cloudops.StoragePoolUpdateResponse, error) {
+	logUpdateRequest(request)
+
+	resp := &cloudops.StoragePoolUpdateResponse{
+		InstanceStorage:     make([]*cloudops.StoragePoolSpec, 0),
+		ResizeOperationType: request.ResizeOperationType,
+	}
+
+	currentCapacity := request.CurrentDriveCount * request.CurrentDriveSize
+	newDeltaCapacity := request.DesiredCapacity - currentCapacity
+	if newDeltaCapacity < 0 {
+		return nil, &cloudops.ErrInvalidStoragePoolUpdateRequest{
+			Request: request,
+			Reason: fmt.Sprintf("reducing instance storage capacity is not supported"+
+				"current: %d GiB requested: %d GiB", currentCapacity, request.DesiredCapacity),
+		}
+	}
+
+	if newDeltaCapacity == 0 {
+		return nil, cloudops.ErrCurrentCapacityEqualToRequested
+	}
+
+	if request.CurrentDriveCount > 0 && len(request.CurrentDriveType) == 0 {
+		return nil, &cloudops.ErrInvalidStoragePoolUpdateRequest{
+			Request: request,
+			Reason: fmt.Sprintf("for storage update operation, current drive" +
+				"type is required to be provided if drives already exist"),
+		}
+	}
+
+	logrus.Debugf("instance currently has %d X %d GiB %s drives",
+		request.CurrentDriveCount, request.CurrentDriveSize, request.CurrentDriveType)
+	filteredRows := filterDecisionMatrix(decisionMatrix, request.CurrentIOPS, request.CurrentDriveType)
+
+ROW_LOOP:
+	for _, row := range filteredRows {
+		switch request.ResizeOperationType {
+		case api.SdkStoragePool_RESIZE_TYPE_ADD_DISK:
+			// Add drives equivalent to newDeltaCapacity
+			logrus.Debugf("check if we can add drive(s) for atleast: %d GiB", newDeltaCapacity)
+			instStorage, err := instanceStorageForRow(row, newDeltaCapacity, request.CurrentDriveSize)
+			if err != nil {
+				if err == cloudops.ErrStorageDistributionCandidateNotFound {
+					continue ROW_LOOP
+				}
+
+				return nil, err
+			}
+
+			resp.InstanceStorage = append(resp.InstanceStorage, instStorage)
+			return resp, nil
+		case api.SdkStoragePool_RESIZE_TYPE_RESIZE_DISK:
+			// Resize existing drives equivalent to newDeltaCapacity
+			if request.CurrentDriveCount == 0 {
+				return nil, &cloudops.ErrInvalidStoragePoolUpdateRequest{
+					Request: request,
+					Reason: fmt.Sprintf("requested resize operation type cannot be " +
+						"accomplished as no existing drives were provided"),
+				}
+			}
+
+			newMinDeltaCapacityPerDrive := newDeltaCapacity / request.CurrentDriveCount
+
+			logrus.Debugf("min delta capacity per drive needed: %d GiB num drives: %d",
+				newMinDeltaCapacityPerDrive, request.CurrentDriveCount)
+
+			// check in matrix if this drive can be resized to newMinCapacityPerDrive
+			newMinCapacityPerDrive := request.CurrentDriveSize + newMinDeltaCapacityPerDrive
+			logrus.Debugf("need to resize drive to atleast: %d GiB", newMinCapacityPerDrive)
+			instStorage, err := instanceStorageForRow(row, newMinCapacityPerDrive, 0)
```

Comment:
instanceStorageForRow also increments the DriveCount, but this switch case only wants to resize the disk. 
Probably we don't need to check the matrix, but just find the row for the current instStorage and check if resizing the drives is within row.MaxSize.